test/gradcheck.jl

```diff
@@ -1129,8 +1132,40 @@ end
     end
 
     @testset "unary pairwise" begin
-      @test gradtest(X->pairwise(metric, X; dims=2), randn(rng, D, P))
-      @test gradtest(Xt->pairwise(metric, Xt; dims=1), randn(rng, P, D))
+
+      X = randn(D, P)
+      Xt = randn(P, D)
+      @test gradtest(X->pairwise(metric, X; dims=2), X)
+      @test gradtest(Xt->pairwise(metric, Xt; dims=1), Xt)
+
+      @testset "X contains duplicates" begin
+        # Zygote's gradtest isn't sufficiently accurate to assess this, so we use
+        # FiniteDifferences.jl instead.
+        Y = hcat(X, X)
+
+        Δ = randn(size(Y, 2), size(Y, 2))
+        Δ_fd = FiniteDifferences.j′vp(
+          central_fdm(5, 1), X -> pairwise(metric, X; dims=2), Δ, Y,
+        )
+        _, pb = Zygote.pullback(X -> pairwise(metric, X; dims=2), Y)
+
+        # This is impressively inaccurate, but at least it doesn't produce a NaN.
+        @test first(Δ_fd) ≈ first(pb(Δ)) atol=1e-9 rtol=1e-9
+      end
+      @testset "Xt contains duplicates" begin
+        # Zygote's gradtest isn't sufficiently accurate to assess this, so we use
+        # FiniteDifferences.jl instead.
+        Y = vcat(Xt, Xt)
+
+        Δ = randn(size(Y, 1), size(Y, 1))
+        Δ_fd = FiniteDifferences.j′vp(
+          central_fdm(5, 1), X -> pairwise(metric, X; dims=1), Δ, Y,
+        )
+        _, pb = Zygote.pullback(X -> pairwise(metric, X; dims=1), Y)
+
+        # This is impressively inaccurate, but at least it doesn't produce a NaN.
```

Comment:
what is this comment about (and the one above) still exist, isn't `1e-9` pretty good accuracy?

Reply:
😂  yes, that should be going